src/uct/ib/base/ib_md.c

```diff
@@ -295,38 +295,44 @@ void *uct_ib_md_mem_handle_thread_func(void *arg)
     while (ctx->length > 0) {
         if (ctx->params != NULL) {
             status = uct_ib_reg_mr(ctx->md, ctx->address, length, ctx->params,
-                                   ctx->access_flags, NULL,
-                                   &ctx->mrs[ctx->mr_idx]);
+                                   ctx->access_flags, NULL, &ctx->mrs[mr_idx]);
             if (status != UCS_OK) {
                 goto err_dereg;
             }
 
         } else {
-            status = uct_ib_dereg_mr(ctx->mrs[ctx->mr_idx]);
+            status = uct_ib_dereg_mr(ctx->mrs[mr_idx]);
             if (status != UCS_OK) {
-                ucs_warn("failed to deregister mr_idx=%d", ctx->mr_idx);
+                ucs_warn("failed to deregister mr_idx=%d", mr_idx);
             }
         }
         ctx->address = UCS_PTR_BYTE_OFFSET(ctx->address, length);
         ctx->length -= length;
         length       = ucs_min(ctx->length, chunk_size);
-        ctx->mr_idx++;
+        mr_idx++;
     }
 
     ucs_trace("%s %p..%p (first_mr_size %zu) took %f usec\n",
               (ctx->params != NULL) ? "reg_mr" : "dereg_mr",
               start, ctx->address, ctx->first_mr_size,
               ucs_time_to_usec(ucs_get_time() - t0));
-    return UCS_STATUS_PTR(UCS_OK);
+    return (void*)(intptr_t)mr_idx;
 
 err_dereg:
-    while (ctx->mr_idx-- > 0) {
-        (void)uct_ib_dereg_mr(ctx->mrs[ctx->mr_idx]);
+    while (mr_idx-- > 0) {
+        (void)uct_ib_dereg_mr(ctx->mrs[mr_idx]);
     }
 
+    ucs_assertv(UCS_STATUS_IS_ERR(status),
+                "Cannot return in-progress memory thread handling");
     return UCS_STATUS_PTR(status);
 }
 
+typedef struct {
```

Comment:
IMO better move it to the beginning of this file

Reply:
fixed